ada/ast.py

```diff
@@ -8554,6 +8561,33 @@ class ObjectDecl(BasicDecl):
     defining_env = Property(Entity.type_expr.defining_env)
     type_expression = Property(Entity.type_expr)
 
+    @langkit_property(public=True, return_type=Bool)
+    def is_constant_object():
+        """
+        Return whether this object is constant or not.
+        """
+        return Or(
+            Self.has_constant.as_bool,
+
+            # A GenericFornalObjDecl is constant if the Mode is `in`
+            (
+                Self.parent.is_a(T.GenericFormalObjDecl)
+                & (Self.mode.is_a(Mode.alt_in, Mode.alt_default))
+            ),
+
+            # Renaming clause is:
+            Entity.renaming_clause.then(
+                lambda rc: If(
+                    # Always constant if it renames a function_call
+                    rc.renamed_object.is_call,
+                    True,
+                    # Constant if the renamed object is constant
+                    rc.renamed_object.referenced_decl
+                    .cast_or_raise(T.ObjectDecl).is_constant_object
```

Comment:
The renamed object could be a parameter as well. I think you should have an abstract `is_constant_object` property on `BasicDecl` so that you can recurse easily here (it will also be easier to use for users).